src/core/congestion_control.h

```diff
@@ -4,84 +4,81 @@
-    //
-    uint8_t Exemptions;
-
-    uint64_t TimeOfLastAck; // millisec
-    uint64_t TimeOfCongAvoidStart; // millisec
-    uint32_t KCubic; // millisec
-    uint32_t PrevKCubic; // millisec
-    uint32_t WindowMax; // bytes
-    uint32_t PrevWindowMax; // bytes
-    uint32_t WindowLastMax; // bytes
-    uint32_t PrevWindowLastMax; // bytes
-
-    //
-    // This variable tracks the largest packet that was outstanding at the time
-    // the last congestion event occurred. An ACK for any packet number greater
-    // than this indicates recovery is over.
-    //
-    uint64_t RecoverySentPacketNumber;
-
+    // Name of congestion control algorithm
+    const char* Name;
+    
+    BOOLEAN (*QuicCongestionControlCanSend)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc
+        );
+
+    void (*QuicCongestionControlSetExemption)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc,
+        _In_ uint8_t NumPackets
+        );
+
+    void (*QuicCongestionControlInitialize)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc,
+        _In_ const QUIC_SETTINGS* Settings
+        );
+
+    void (*QuicCongestionControlReset)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc
+        );
+
+    uint32_t (*QuicCongestionControlGetSendAllowance)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc,
+        _In_ uint64_t TimeSinceLastSend,
+        _In_ BOOLEAN TimeSinceLastSendValid
+        );
+
+    void (*QuicCongestionControlOnDataSent)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc,
+        _In_ uint32_t NumRetransmittableBytes
+        );
+
+    BOOLEAN (*QuicCongestionControlOnDataInvalidated)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc,
+        _In_ uint32_t NumRetransmittableBytes
+        );
+
+    BOOLEAN (*QuicCongestionControlOnDataAcknowledged)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc,
+        _In_ uint64_t TimeNow, // microsecond
+        _In_ uint64_t LargestPacketNumberAcked,
+        _In_ uint32_t NumRetransmittableBytes,
+        _In_ uint32_t SmoothedRtt
+        );
+
+    void (*QuicCongestionControlOnDataLost)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc,
+        _In_ uint64_t LargestPacketNumberLost,
+        _In_ uint64_t LargestPacketNumberSent,
+        _In_ uint32_t NumRetransmittableBytes,
+        _In_ BOOLEAN PersistentCongestion
+        );
+
+    void (*QuicCongestionControlOnSpuriousCongestionEvent)(
+        _In_ struct QUIC_CONGESTION_CONTROL* Cc
+        );
+
+    void (*QuicCongestionControlLogOutFlowStatus)(
+        _In_ const struct QUIC_CONGESTION_CONTROL* Cc
+        );
+
+    uint8_t (*QuicCongestionControlGetExemptions)(
+        _In_ const struct QUIC_CONGESTION_CONTROL* Cc
+        );
+
+    uint32_t (*QuicCongestionControlGetBytesInFlightMax)(
+        _In_ const struct QUIC_CONGESTION_CONTROL* Cc
+        );
+
+    uint64_t Ctx[104 / sizeof(uint64_t)];
 } QUIC_CONGESTION_CONTROL;
 
+#define QUIC_CONGESTION_CONTROL_CONTEXT_SIZE (RTL_FIELD_SIZE(QUIC_CONGESTION_CONTROL, Ctx))
```

Comment:
I'm not a big fan of this mechanism of making space for the various algorithms, but I don't currently have a good suggestion for something better... @maolson-msft ideas? How does TCP do this?

Reply:
I'm wondering, if instead, congestion_control.h should include all the algorithm's headers, and define this as a `max` of their struct sizes.

Reply:
This machanism is coming from [Linux kernel code][1].

This could help us to avoid dynamic memory allocation as we don't have `constexpr` here in C11.

[1]: https://github.com/torvalds/linux/blob/dbe69e43372212527abf48609aba7fc39a6daa27/include/net/inet_connection_sock.h#L137

Reply:
The problem with this is that its not the smallest size possible, and it breaks if we change the algorithms without changing this.